src/node_encoding.cc

```diff
@@ -0,0 +1,196 @@
+using v8::FastApiCallbackOptions;
+using v8::FastApiTypedArray;
+using v8::FastOneByteString;
+using v8::FunctionCallbackInfo;
+using v8::Isolate;
+using v8::Local;
+using v8::MaybeLocal;
+using v8::Object;
+using v8::String;
+using v8::Uint32Array;
+using v8::Uint8Array;
+using v8::Value;
+
+namespace encoding_methods {
+
+static void EncodeUtf8(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  CHECK_GE(args.Length(), 1);
+  CHECK(args[0]->IsString());
+
+  Local<String> str = args[0].As<String>();
+  size_t length = str->Utf8Length(isolate);
+
+  Local<ArrayBuffer> ab;
+  {
+    NoArrayBufferZeroFillScope no_zero_fill_scope(env->isolate_data());
+    std::unique_ptr<BackingStore> bs =
+        ArrayBuffer::NewBackingStore(isolate, length);
+
+    CHECK(bs);
+
+    str->WriteUtf8(isolate,
+                   static_cast<char*>(bs->Data()),
+                   -1,  // We are certain that `data` is sufficiently large
+                   nullptr,
+                   String::NO_NULL_TERMINATION | String::REPLACE_INVALID_UTF8);
+
+    ab = ArrayBuffer::New(isolate, std::move(bs));
+  }
+
+  args.GetReturnValue().Set(Uint8Array::New(ab, 0, length));
+}
+
+static void EncodeIntoUtf8(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  CHECK_GE(args.Length(), 3);
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsUint8Array());
+  CHECK(args[2]->IsUint32Array());
+
+  Local<String> source = args[0].As<String>();
+
+  Local<Uint8Array> dest = args[1].As<Uint8Array>();
+  Local<ArrayBuffer> buf = dest->Buffer();
+  char* write_result = static_cast<char*>(buf->Data()) + dest->ByteOffset();
+  size_t dest_length = dest->ByteLength();
+
+  // results = [ read, written ]
+  Local<Uint32Array> result_arr = args[2].As<Uint32Array>();
+  uint32_t* results = reinterpret_cast<uint32_t*>(
+      static_cast<char*>(result_arr->Buffer()->Data()) +
+      result_arr->ByteOffset());
+
+  int nchars;
+  int written = source->WriteUtf8(
+      isolate,
+      write_result,
+      dest_length,
+      &nchars,
+      String::NO_NULL_TERMINATION | String::REPLACE_INVALID_UTF8);
+  results[0] = nchars;
+  results[1] = written;
+}
+
+static void FastEncodeIntoUtf8(
+    Local<Value> receiver,
+    const FastOneByteString& source,
+    const FastApiTypedArray<uint8_t>& destination,
+    const FastApiTypedArray<uint32_t>& result,
+    FastApiCallbackOptions& options) {  // NOLINT(runtime/references)
+#if defined(NODE_HAVE_I18N_SUPPORT)
+  uint8_t* destination_data;
+  CHECK(destination.getStorageIfAligned(&destination_data));
+
+  uint32_t* results;
+  CHECK(result.getStorageIfAligned(&results));
+
+  size_t destination_length = destination.length();
+  size_t source_length = source.length;
+  size_t written = std::min(source_length, destination_length);
+
+  U8_GET_UNSAFE(source.data, 0, *results);
```

Comment:
This doesn’t do the right thing at all. `U8_GET_UNSAFE()` expects UTF-8 data as its first argument; it receives Latin-1 data here. It also writes to its third argument, which is being overwritten in the next line here, so this line is essentially effectless. And finally, it only reads a single codepoint at the beginning of the string, not the entire string.